File: Scripts/invert.py
# Invert all cell states in the current selection.
# Author: Andrew Trevorrow (andrew@trevorrow.com), June 2006.
# Updated to use exit command, Nov 2006.

from glife import rect
from time import time
import golly as g

r = rect( g.getselrect() )

if r.empty: g.exit("There is no selection.")

oldsecs = time()
for row in xrange(r.top, r.top + r.height):
   # when inverting large selections it's nicer to
   # give some indication of progress
   newsecs = time()
   if newsecs - oldsecs >= 1.0:
      oldsecs = newsecs
      g.update()

   # also allow keyboard interaction
   g.dokey( g.getkey() )

   for col in xrange(r.left, r.left + r.width):
      g.setcell(col, row, 1 - g.getcell(col, row))

if not r.visible(): g.fitsel()

File: Scripts/Lua/invert.lua
-- Invert all cell states in the current selection.
-- Author: Andrew Trevorrow (andrew@trevorrow.com), Mar 2016.

local g = golly()
local gp = require "gplus"

-- re-assigning inner loop functions results in a 10% speed up
local setcell = g.setcell
local getcell = g.getcell

local r = gp.rect(g.getselrect())
if r.empty then g.exit("There is no selection.") end

-- local t1 = os.clock()

local oldsecs = os.clock()
local maxstate = g.numstates() - 1

for row = r.top, r.bottom do
    -- if large selection then give some indication of progress
    local newsecs = os.clock()
    if newsecs - oldsecs >= 1.0 then
        oldsecs = newsecs
        g.update()
    end
    for col = r.left, r.right do
        setcell(col, row, maxstate - getcell(col, row))
    end
end

if not r.visible() then g.fitsel() end

-- g.show(""..(os.clock()-t1))
